src/waypoint.jsx

```diff
@@ -23,31 +23,70 @@ function debugLog() {
   console.log(arguments); // eslint-disable-line no-console
 }
 
-const resizeEventHandlers = {
-  size: 0,
-  index: 0,
-};
+class TargetEventHandlers {
+  constructor(target) {
+    this.target = target;
+    this.events = {};
+  }
 
-function handleResize(event) {
-  resizeEventHandlers.forEach(handler => handler(event));
-}
+  getEventHandlers(eventName) {
+    if (!this.events[eventName]) {
+      this.events[eventName] = {
+        size: 0,
+        index: 0,
+        handlers: {},
+      };
+    }
+    return this.events[eventName];
+  }
+
+  handleEvent(eventName, event) {
+    const { handlers } = this.getEventHandlers(eventName);
+    Object.keys(handlers).forEach(function(index) {
```

Comment:
It seems to be pretty good in the Chrome timeline. Do you have any ideas for something faster? I guess we could replace the .forEach with a plain old for loop, but other than that I don't really see anything.